OSDOCS-8395: Migrated the Scaling lab from the ROSA workshop (#74502)
openshift/openshift-docs: cloud_experts_tutorials/cloud-experts-deploying-application/cloud-experts-deploying-application-scaling.adoc

:_mod-docs-content-type: ASSEMBLY
[id="cloud-experts-deploying-application-scaling"]
= Tutorial: Scaling an application
include::_attributes/attributes-openshift-dedicated.adoc[]
:context: cloud-experts-deploying-application-scaling
:source-highlighter: pygments
:pygments-style: emacs
:icons: font

toc::[]

//rosaworkshop.io content metadata
//Brought into ROSA product docs 2024-04-10

== Scaling

You can manually scale your pods or you can automatically scale your pods by using the Horizontal Pod Autoscaler (HPA). You can also scale your cluster nodes.

=== Manual pod scaling

> You can manually scale the pods and nodes of your applications:

Reviewer: the title of this section is pod scaling.  So not sure if "nodes" should be here.

> * by changing your replicaset and deployment definition 

Reviewer: it would be more accurate to say "...replicaset _or_ deployment definition". As you don't have to modify both as youd generally be using one or the other

* by using the command line
* by using the web console

> * by changing your *deployment* definition in the `ostoy-frontend-deployment.yaml`

Reviewer: this is the same as the first bullet

> This workshop starts by using only one pod for the microservice. By using one pod, the Kubernetes Replication Controller strives to keep one pod alive. You can also define pod autoscaling by using the link:https://docs.openshift.com/container-platform/latest/nodes/pods/nodes-pods-autoscaling.html[Horizontal Pod Autoscaler](HPA). The HPA is based on workload, and it expands past defined limits if needed.

Reviewer: In the second sentence I'd start with "By defining a replica of "1" in our deployment definition, the Kubernetes...".  The two idea's here don't connect well when I am reading it.  It is not clear to me how we went from saying we only defined 1 pod and how the HPA is connected.  The second part seems disconnected.  Can we try to connect the ideas?  Something like "You can also define pod autoscaling by using the Horizontal Pod Autoscaler (HPA) which is based on the load and will scale out more pods, beyond our initial definition, if high load is experienced."

.Prerequisites

* Create a cluster with the xref:../../cloud_experts_tutorials/cloud-experts-deploying-application/cloud-experts-deploying-application-lab-overview.adoc[OSToy microservice deployed].

> * Download the link:https://www.rosaworkshop.io/ostoy/yaml/ostoy-microservice-deployment.yaml[ostoy-microservice-deployment.yaml].

Reviewer: the microservice deployment manifest is not needed to be downloaded.  What is the prerequisites would be:
1. A ROSA Cluster
2. The OSTOY application successfully deployed.

> .Procedure 

Reviewer: This section is missing accomplishing the procedure for updating the deployment manifest.

. In the OSToy app, click the *Networking* tab in the left menu.
. In the "Intra-cluster Communication" section, there is a box that randomly changes colors. Inside the box, you see the microservice's pod name. There is only one box shown because there is only one microservice pod. 
+
image::deploy-scale-network.png[HPA Menu]
+ 

> Now, scale up the microservice pods and observe the number of boxes change.

Reviewer: Remove this sentence.  The next step is not doing the scaling but rather further verification that only 1 pod exists.

> . Run the following command to confirm that there is only one pod running on the microservice:

Reviewer: "..._for_ the microservice"

+
[source,terminal,highlight='4']
----
$ oc get pods
----
+
.Example output
[source,terminal]
----
NAME                                  READY     STATUS    RESTARTS   AGE
ostoy-frontend-679cb85695-5cn7x       1/1       Running   0          1h
ostoy-microservice-86b4c6f559-p594d   1/1       Running   0          1h
----

. Now, scale the application by using the CLI or using the Web UI:
+

> ** Run the following command increases the number of pods from `1` to `3`:

Reviewer: Do you mean "Run the following command to increase the number of pods from 1 to 3"?

+
[source,terminal]
----
$ oc scale deployment ostoy-microservice --replicas=3
----
+

> ** In the OSToy application, from the left menu, click *Workloads > Deployments > ostoy-microservice*.  On the left side of the page, you see a blue circle with a "1 Pod" label in the middle. Selecting the arrows next to this scales the number of pods. Select the up arrow to `3`.

Reviewer: This is accomplished from the OCP web console, not through the demo application.  Something to consider.  The previous command asks the reader to increase the pods to 3 via the cli.  They will now see "3 pods" in the circle.  Do you mean to tell them to scale to 4 pods?  If so then you need a new screen shot.  Also, in the original workshop we went up and then down with 3 pods being the max as more is not really needed to convey the concept.

+

> image::deploy-scale-uiscale.png[UI Scale]

Reviewer: Depending on what you want the flow to be this screenshot might need updating.  Are there now 4 pods, or 2?  Also, it might be helpful to the reader to pinpoint where they should be looking to see the correct number as there are 2 lines in this screenshot.

.Verification

Check your pod counts with the following methods:

* In the web UI, select *Workloads > Deployments > ostoy-microservice*.
+
image::deploy-scale-verify-workload.png[Verify the workload pods]

* From the CLI, run the follow command to confirm that you are using three pods:
+
[source,terminal]
----
$ oc get pods
----
+
.Example output
[source,terminal]
----
NAME                                  READY     STATUS    RESTARTS   AGE
ostoy-frontend-679cb85695-5cn7x       1/1       Running   0          1h
ostoy-microservice-86b4c6f559-p594d   3/3       Running   0          1h
----

* You can also confirm that there are three pods being used by selecting **Networking** in the left menu of the OSToy app. There should be three colored boxes for the three pods.
+
image::deploy-scale-colorspods.png[UI Scale]

=== Pod Autoscaling

> {product-title} offers a link:https://docs.openshift.com/container-platform/latest/nodes/pods/nodes-pods-autoscaling.html[Horizontal Pod Autoscaler] (HPA). The horizontal pod autoscaler (HPA) uses metrics to decide if increase or decrease the scale your cluster's worker pods.

Reviewer: The second sentence needs updating.  I think something to the effect of "The HPA uses metrics to decide whether to increase of decrease the number of pods."

.Procedure

. In the web UI, select *Auto Scaling* in the left menu.
+

> image::deploy-scale-hpa-menu.png[HPA Menu]

Reviewer: The menu item changed slightly to "Pod Auto Scaling".  Should we consider a new screenshot. (attached)
<img width="226" alt="Screenshot 2024-08-06 at 4 48 09 PM" src="https://github.com/user-attachments/assets/23fc4344-d2c9-4a04-aecb-301a2e71cd13">

. Create the Horizontal Pod Autoscaler by running the following command:
+
[source,terminal]
----
$ oc autoscale deployment/ostoy-microservice --cpu-percent=80 --min=1 --max=10
----
+
This command creates an HPA that maintains between 1 and 10 replicas of the pods controlled by the *ostoy-microservice* deployment. HPA increases and decreases the number of replicas, through the deployment, to keep an average CPU use across all pods of 80%. There is an average CPU usage of 40 millicores since each pod requests 50 millicores.

> . On the *Auto Scaling* page, select the *Increase the load* button. 

Reviewer: The page title is "Horizontal Pod Autoscaling"

+
[IMPORTANT]
====
This action generates some CPU intensive calculations, the page can become unresponsive. This is an expected response. Click the *Increase the Load* button only once. For more information about the process, see the link:https://github.com/openshift-cs/ostoy/blob/master/microservice/app.js#L32[microservice's GitHub repository]. 
====
+
After a few minutes, the new pods display on the page, as represented by colored rectangles.
+
[NOTE]
====
The page can experience a bit of lag.
====

.Verification

Check your pod counts with the following methods:

* In the OSToy application's web UI, see the remote pods box:
+
image::deploy-scale-hpa-mainpage.png[HPA Main]
+

> Since there is only one pod, increasing the workload should trigger an increase of nodes.

Reviewer: increase of "pods", not nodes in this section.

:_mod-docs-content-type: ASSEMBLY
[id="cloud-experts-deploying-application-scaling"]
= Tutorial: Scaling an application
include::_attributes/attributes-openshift-dedicated.adoc[]
:context: cloud-experts-deploying-application-scaling
:source-highlighter: pygments
:pygments-style: emacs
:icons: font

toc::[]

//rosaworkshop.io content metadata
//Brought into ROSA product docs 2024-04-10

== Scaling

> You can manually scale your pods or you can automatically scale your pods by using the Horizontal Pod Autoscaler (HPA). You can also scale your cluster nodes.

Author: @0kashi can you address this one?

> You can manually scale your pods or you can automatically scale your pods by using the Horizontal Pod Autoscaler (HPA). You can also scale your cluster nodes.

Reviewer: Are you asking if you can have an HPA but then manually scale the pods?  If so, then you can.

+
image::deploy-scale-network.png[HPA Menu]
+ 

. Run the following command to confirm that there is only one pod running for the microservice:
+
[source,terminal,highlight='4']
----
$ oc get pods
----
+
.Example output
[source,terminal]
----
NAME                                  READY     STATUS    RESTARTS   AGE
ostoy-frontend-679cb85695-5cn7x       1/1       Running   0          1h
ostoy-microservice-86b4c6f559-p594d   1/1       Running   0          1h
----

. Download the link:https://www.rosaworkshop.io/ostoy/yaml/ostoy-microservice-deployment.yaml[ostoy-microservice-deployment.yaml] and save it to your local machine.
. Change the deployment definition to three pods instead of one, similar to the following example:
+
[source,yaml]
----
spec:
    selector:
      matchLabels:
        app: ostoy-microservice
    replicas: 3
----

. Now, scale the application by using the CLI or using the Web UI:
+
** Run the following command to increase the number of pods from `1` to `3`:
+
[source,terminal]
----
$ oc scale deployment ostoy-microservice --replicas=3
----
+

> ** In the OpenShift web console, from the left menu, click *Workloads > Deployments > ostoy-microservice*.  On the left side of the page, you see a blue circle with a "1 Pod" label in the middle. Selecting the arrows next to this scales the number of pods. Select the up arrow to `3`.

Reviewer: If we followed this flow we would already be at 3 pods and clicking twice on that arrow would give us 5 pods.  We need to look at the flow as a whole.

.Procedure

. In the web UI, select *Pod Auto Scaling* in the left menu.
+
image::deploy-scale-hpa-menu.png[HPA Menu]

. Create the Horizontal Pod Autoscaler by running the following command:
+
[source,terminal]
----
$ oc autoscale deployment/ostoy-microservice --cpu-percent=80 --min=1 --max=10
----
+
This command creates an HPA that maintains between 1 and 10 replicas of the pods controlled by the *ostoy-microservice* deployment. HPA increases and decreases the number of replicas, through the deployment, to keep an average CPU use across all pods of 80%. There is an average CPU usage of 40 millicores since each pod requests 50 millicores.

. On the *Horizontal Pod Autoscaling* page, select the *Increase the load* button. 
+
[IMPORTANT]
====
This action generates some CPU intensive calculations, the page can become unresponsive. This is an expected response. Click the *Increase the Load* button only once. For more information about the process, see the link:https://github.com/openshift-cs/ostoy/blob/master/microservice/app.js#L32[microservice's GitHub repository]. 
====
+

> After a few minutes, the new pods display on the page, as represented by colored rectangles.

Reviewer 2: ```suggestion
After a few minutes, the new pods display on the page, as represented by colored boxes.
```

Reviewer 2: Changing to "boxes" for consistency purposes.

.Verification

Check your pod counts with the following methods:

* In the OSToy application's web UI, see the remote pods box:
+
image::deploy-scale-hpa-mainpage.png[HPA Main]
+
Since there is only one pod, increasing the workload should trigger an increase of pods.
+
* In the CLI, run the following command:
+
[source,terminal]
----
oc get pods --field-selector=status.phase=Running | grep microservice
----
+
.Example output
+
[source,terminal]
----
ostoy-microservice-79894f6945-cdmbd   1/1     Running   0          3m14s
ostoy-microservice-79894f6945-mgwk7   1/1     Running   0          4h24m
ostoy-microservice-79894f6945-q925d   1/1     Running   0          3m14s
----

* You can also verify autoscaling from the {cluster-manager} 
+
. In the OpenShift web console left menu, click *Observe > Dashboards*.
. In the dashboard, select *Kubernetes / Compute Resources / Namespace (Pods)* and your namespace *ostoy*.
+
image::deploy-scale-hpa-metrics.png[Select metrics]
+
. A graph appears showing your resource usage across CPU and memory. The top graph will show recent CPU consumption per pod and the lower graph indicates memory usage. The following lists the call outs in the graph: 

> .. As soon as the load started to increase (A), 

Reviewer 2: ```suggestion
.. The load started to increase (A), 
```

> .. two new pods started to spin up (B, C). 

Reviewer 2: ```suggestion
.. Two new pods started to spin up (B, C). 
```

> .. The thickness of each graph is its CPU consumption indicating which pods handled more load. 

Reviewer 2: ```suggestion
.. The thickness of each graph is its CPU consumption, indicating which pods handled more load. 
```

.. You can also see that the load decreased (D), after which, the pods were spun back down.
+
image::deploy-scale-metrics.png[Select metrics]

=== Node Autoscaling

> {product-title} allows you to use link:https://docs.openshift.com/rosa/rosa_cluster_admin/rosa_nodes/rosa-nodes-about-autoscaling-nodes.html[node autoscaling]. In this scenario, you will create a new project with a job that has a large workload that this cluster cannot handle. With autoscaling enabled, when your load is larger than you can currently handle, it should force the cluster to automatically create new nodes to handle the load.

Reviewer 2: ```suggestion
{product-title} allows you to use link:https://docs.openshift.com/rosa/rosa_cluster_admin/rosa_nodes/rosa-nodes-about-autoscaling-nodes.html[node autoscaling]. In this scenario, you will create a new project with a job that has a large workload that the cluster cannot handle. With autoscaling enabled, when your load is larger than you can currently handle, it should force the cluster to automatically create new nodes to handle the load.
```

ostoy-frontend-679cb85695-5cn7x       1/1       Running   0          1h
ostoy-microservice-86b4c6f559-p594d   1/1       Running   0          1h
----

. Download the link:https://www.rosaworkshop.io/ostoy/yaml/ostoy-microservice-deployment.yaml[ostoy-microservice-deployment.yaml] and save it to your local machine.
. Change the deployment definition to three pods instead of one, similar to the following example:
+
[source,yaml]
----
spec:
    selector:
      matchLabels:
        app: ostoy-microservice
    replicas: 3
----

. Apply the replica changes by running the following command:
+
[source,terminal]
----
$ oc apply -f ostoy-microservice-deployment.yaml
----
+
[NOTE]
====
You can also edit the `ostoy-microservice-deployment.yaml` file in the OpenShift Web Console by going to *Workloads > Deployments > ostoy-microservice > YAML* tab.
====

. Confirm that there are now 3 pods by running the following command:
+
[source,terminal]
----
$ oc get pods
----
+
The output shows that there are now 3 pods instead of only one.
+
.Example output
+
[source,terminal]
----
NAME                                  READY   STATUS    RESTARTS   AGE

> ostoy-frontend-5fbcc7d9-rzlgz         1/1     Running   0          26m

Reviewer: no. It should match the expectation.  The flow of this lab should be given another once over so that the scale bia each of the methods are linear.

ostoy-microservice-6666dcf455-2lcv4   1/1     Running   0          81s
ostoy-microservice-6666dcf455-5z56w   1/1     Running   0          81s
ostoy-microservice-6666dcf455-tqzmn   1/1     Running   0          26m
----

. Now, scale the application by using the CLI or using the Web UI:
+
** Run the following command to increase the number of pods from `3` to `2`:
+
[source,terminal]
----
$ oc scale deployment ostoy-microservice --replicas=2
----
+
** In the OpenShift web console, from the left menu, click *Workloads > Deployments > ostoy-microservice*.  On the left side of the page, you see a blue circle with a "3 Pod" label in the middle. Selecting the arrows next to this scales the number of pods. Select the down arrow to `2`.
+
image::deploy-scale-uiscale.png[UI Scale]

.Verification

Check your pod counts with the following methods:

* In the web UI, select *Workloads > Deployments > ostoy-microservice*.
+
image::deploy-scale-verify-workload.png[Verify the workload pods]

> * From the CLI, run the follow command to confirm that you are using three pods:

Reviewer 3: ```suggestion
* From the CLI, confirm that you are using three pods by running the following command: 
```

+
[source,terminal]
----
$ oc get pods
----
+
.Example output
[source,terminal]
----
NAME                                  READY   STATUS    RESTARTS   AGE
ostoy-frontend-5fbcc7d9-rzlgz         1/1     Running   0          75m
ostoy-microservice-6666dcf455-2lcv4   1/1     Running   0          50m
ostoy-microservice-6666dcf455-tqzmn   1/1     Running   0          75m
----

> * You can also confirm that there are three pods being used by selecting **Networking** in the left menu of the OSToy app. There should be two colored boxes for the two pods.

Reviewer 3: Same feedback (eg "left menu") here.

+
image::deploy-scale-colorspods.png[UI Scale]

=== Pod Autoscaling

> {product-title} offers a link:https://docs.openshift.com/container-platform/latest/nodes/pods/nodes-pods-autoscaling.html[Horizontal Pod Autoscaler] (HPA). The HPA uses metrics to decide if increase or decrease the number of pods.

Reviewer 3: Might want to revise the second sentence for clarity. eg

```suggestion
{product-title} offers a link:https://docs.openshift.com/container-platform/latest/nodes/pods/nodes-pods-autoscaling.html[Horizontal Pod Autoscaler] (HPA). The HPA uses metrics to decide if increasing or decreasing the number of pods is necessary.
```

.Procedure

> . In the web UI, select *Pod Auto Scaling* in the left menu.

Reviewer 3: If opting to remove "sensory elements" maybe consider

```suggestion
. In the web UI, select *Pod Auto Scaling* in the navigation panel.
```

----
+
This command creates an HPA that maintains between 1 and 10 replicas of the pods controlled by the *ostoy-microservice* deployment. HPA increases and decreases the number of replicas, through the deployment, to keep an average CPU use across all pods of 80%. There is an average CPU usage of 40 millicores since each pod requests 50 millicores.

. On the *Pod Auto Scaling > Horizontal Pod Autoscaling* page, select the *Increase the load* button. 
+
[IMPORTANT]
====
Because increasing the load generates some CPU intensive calculations, the page can become unresponsive. This is an expected response. Click the *Increase the Load* button only once. For more information about the process, see the link:https://github.com/openshift-cs/ostoy/blob/master/microservice/app.js#L32[microservice's GitHub repository]. 
====
+
After a few minutes, the new pods display on the page, as represented by colored boxes.
+
[NOTE]
====
The page can experience a bit of lag.
====

.Verification

Check your pod counts with the following methods:

* In the OSToy application's web UI, see the remote pods box:
+
image::deploy-scale-hpa-mainpage.png[HPA Main]
+

> Since there is only one pod, increasing the workload should trigger an increase of pods.

Reviewer 3: Suggest "because" 

https://www.ibm.com/docs/en/ibm-style?topic=word-usage#since

+
* In the CLI, run the following command:
+
[source,terminal]
----
oc get pods --field-selector=status.phase=Running | grep microservice
----
+
.Example output
+
[source,terminal]
----
ostoy-microservice-79894f6945-cdmbd   1/1     Running   0          3m14s
ostoy-microservice-79894f6945-mgwk7   1/1     Running   0          4h24m
ostoy-microservice-79894f6945-q925d   1/1     Running   0          3m14s
----

* You can also verify autoscaling from the {cluster-manager} 
+
. In the OpenShift web console left menu, click *Observe > Dashboards*.
. In the dashboard, select *Kubernetes / Compute Resources / Namespace (Pods)* and your namespace *ostoy*.
+
image::deploy-scale-hpa-metrics.png[Select metrics]
+

> . A graph appears showing your resource usage across CPU and memory. The top graph will show recent CPU consumption per pod and the lower graph indicates memory usage. The following lists the call outs in the graph: 

Reviewer 3: Suggest removal of "will"

```suggestion
. A graph appears showing your resource usage across CPU and memory. The top graph shows recent CPU consumption per pod and the lower graph indicates memory usage. The following lists the call outs in the graph: 
```

> . A graph appears showing your resource usage across CPU and memory. The top graph shows recent CPU consumption per pod and the lower graph indicates memory usage. The following lists the call outs in the graph: 

Reviewer 4: ```suggestion
. A graph appears showing your resource usage across CPU and memory. The top graph shows recent CPU consumption per pod and the lower graph indicates memory usage. The following lists the callouts in the graph: 
```

> .. The load started to increase (A), 

Reviewer 4: ```suggestion
.. The load increased (A).
```

> .. Two new pods started to spin up (B, C). 

Reviewer 4: ```suggestion
.. Two new pods were created (B and C). 
```

> .. The thickness of each graph is its CPU consumption, indicating which pods handled more load. 

Reviewer 4: ```suggestion
.. The thickness of each graph represents the CPU consumption and indicates which pods handled more load. 
```

> .. You can also see that the load decreased (D), after which, the pods were spun back down.

Reviewer 4: ```suggestion
.. The load decreased (D), and the pods were deleted.
```

+
image::deploy-scale-metrics.png[Select metrics]

=== Node Autoscaling

> {product-title} allows you to use link:https://docs.openshift.com/rosa/rosa_cluster_admin/rosa_nodes/rosa-nodes-about-autoscaling-nodes.html[node autoscaling]. In this scenario, you will create a new project with a job that has a large workload that the cluster cannot handle. With autoscaling enabled, when your load is larger than you can currently handle, it should force the cluster to automatically create new nodes to handle the load.

Reviewer 4: ```suggestion
{product-title} allows you to use link:https://docs.openshift.com/rosa/rosa_cluster_admin/rosa_nodes/rosa-nodes-about-autoscaling-nodes.html[node autoscaling]. In this scenario, you will create a new project with a job that has a large workload that the cluster cannot handle. With autoscaling enabled, when the load is larger than your current capacity, the cluster will automatically create new nodes to handle the load.
```

.Prerequisites

* Autoscaling is enabled on your machine pools.

.Procedure

. Create a new project called `autoscale-ex` by running the following command:
+
[source,terminal]
----
$ oc new-project autoscale-ex
----

> . Create the job with the following command:

Reviewer 3: Small nit

```suggestion
. Create the job by running the following command:
```

+
[source,terminal]
----
$ oc create -f https://raw.githubusercontent.com/openshift-cs/rosaworkshop/master/rosa-workshop/ostoy/yaml/job-work-queue.yaml
----
+

> . Allow the pods some time to create and then, run the following command to see the pods:

Reviewer 4: ```suggestion
. After a few minutes, run the following command to see the pods:
```

+
[source,terminal]
----
$ oc get pods
----
+
.Example output
+
[source,terminal]
----
NAME                     READY   STATUS    RESTARTS   AGE
work-queue-5x2nq-24xxn   0/1     Pending   0          10s
work-queue-5x2nq-57zpt   0/1     Pending   0          10s
work-queue-5x2nq-58bvs   0/1     Pending   0          10s
work-queue-5x2nq-6c5tl   1/1     Running   0          10s
work-queue-5x2nq-7b84p   0/1     Pending   0          10s
work-queue-5x2nq-7hktm   0/1     Pending   0          10s
work-queue-5x2nq-7md52   0/1     Pending   0          10s
work-queue-5x2nq-7qgmp   0/1     Pending   0          10s
work-queue-5x2nq-8279r   0/1     Pending   0          10s
work-queue-5x2nq-8rkj2   0/1     Pending   0          10s
work-queue-5x2nq-96cdl   0/1     Pending   0          10s
work-queue-5x2nq-96tfr   0/1     Pending   0          10s
----

> . Because there are many pods in a `Pending` state, this status should trigger the autoscaler to create more nodes in your machine pool. Allow more time to pass to create these worker nodes.

Reviewer 4: ```suggestion
. Because there are many pods in a `Pending` state, this status should trigger the autoscaler to create more nodes in your machine pool. Allow time to create these worker nodes.
```

ip-10-0-153-68.us-west-2.compute.internal    Ready    worker         2m12s   v1.23.5+3afdacb
ip-10-0-165-183.us-west-2.compute.internal   Ready    worker         2m8s    v1.23.5+3afdacb
ip-10-0-176-123.us-west-2.compute.internal   Ready    infra,worker   22h     v1.23.5+3afdacb
ip-10-0-195-210.us-west-2.compute.internal   Ready    master         23h     v1.23.5+3afdacb
ip-10-0-196-84.us-west-2.compute.internal    Ready    master         23h     v1.23.5+3afdacb
ip-10-0-203-104.us-west-2.compute.internal   Ready    worker         2m6s    v1.23.5+3afdacb
ip-10-0-217-202.us-west-2.compute.internal   Ready    master         23h     v1.23.5+3afdacb
ip-10-0-225-141.us-west-2.compute.internal   Ready    worker         23h     v1.23.5+3afdacb
ip-10-0-231-245.us-west-2.compute.internal   Ready    worker         2m11s   v1.23.5+3afdacb
ip-10-0-245-27.us-west-2.compute.internal    Ready    worker         2m8s    v1.23.5+3afdacb
ip-10-0-245-7.us-west-2.compute.internal     Ready    worker         23h     v1.23.5+3afdacb
----
+
You can see the worker nodes were automatically created to handle the workload.

> . Use the following command to return to the OSToy app

Reviewer 3: Suggest swapping order:

```suggestion
. Return to the OSToy app by entering the following command:
```
